amqp_job_queue.go

```diff
@@ -42,24 +43,32 @@ func NewAMQPJobQueue(conn *amqp.Connection, queue string, stateUpdatePoolSize in
 		return nil, err
 	}
 
-	_, err = channel.QueueDeclare("reporting.jobs.builds", true, false, false, false, nil)
-	if err != nil {
-		return nil, err
-	}
-
 	err = channel.ExchangeDeclare("reporting", "topic", true, false, false, false, nil)
 	if err != nil {
 		return nil, err
 	}
 
-	_, err = channel.QueueDeclare("reporting.jobs.logs", true, false, false, false, nil)
+	_, err = channel.QueueDeclare("reporting.jobs.builds", true, false, false, false, nil)
 	if err != nil {
 		return nil, err
 	}
 
-	err = channel.QueueBind("reporting.jobs.logs", "reporting.jobs.logs", "reporting", false, nil)
-	if err != nil {
-		return nil, err
+	if sharded {
+		// This exchange should be declared as sharded using a policy that matches its name.
+		err = channel.ExchangeDeclare("reporting.jobs.logs_sharded", "direct", true, false, false, false, nil)
```

Comment:
I believe we may need to use a `x-modulus-hash` exchange instead of a direct exchange, [as suggested by rabbitmq-sharding](https://github.com/rabbitmq/rabbitmq-sharding#messages-distribution-between-shards-partitioning).

The reason is kind of subtle and it was frankly hard to find in the rabbit docs. In [the routing tutorial](https://www.rabbitmq.com/tutorials/tutorial-four-python.html) under "Multiple bindings", they say:

> It is perfectly legal to bind multiple queues with the same binding key. In our example we could add a binding between X and Q1 with binding key `black`. In that case, the direct exchange will behave like fanout and will broadcast the message to all the matching queues.

So I believe that if we use a direct exchange with the sharding plugin, the same message will get copied to each of the shards, instead of distributed between them.

That's what I saw in a test I ran, but maybe you can see if you can verify that behaviour. 👍 

Side-note: If we _do_ go with the `x-modulus-hash` exchange, we'll need to define a routing key when publishing, which will used for hash partitioning.

Reply:
Yep, this is indeed the case. Thank you for the suggestion!